app/controllers/omniauth_callbacks_controller.rb

```diff
@@ -0,0 +1,25 @@
+class OmniauthCallbacksController < Devise::OmniauthCallbacksController
+  def github
+    @user = User.from_omniauth(request.env['omniauth.auth'])
+
+    if @user.persisted?
+      sign_in_and_redirect @user, event: :authentication
+      flash[:notice] = I18n.t('devise.omniauth_callbacks.success', kind: 'GitHub')
+    else
+      session['devise.github_data'] = request.env['omniauth.auth']
+      redirect_to new_user_registration_url
+    end
+  end
+
+  def google_oauth2
+    @user = User.from_omniauth(request.env['omniauth.auth'])
+
+    if @user.persisted?
+      sign_in_and_redirect @user, event: :authentication
+      flash[:notice] = I18n.t('devise.omniauth_callbacks.success', kind: 'Google')
+    else
+      session['devise.google_data'] = request.env['omniauth.auth']
+      redirect_to new_user_registration_url
+    end
+  end
+end
```

Comment:
Let's get rid of using a local variable in the actions and dry them a bit.
```ruby
private

def user
  @user ||= User.from_omniauth(request.env['omniauth.auth'])  
end
```
Then you can go with `user` property for both actions